src/main/scala/org/kframework/minikore/interfaces/Interfaces.scala

```diff
@@ -0,0 +1,493 @@
+  }
+
+
+  /**
+    * A Node with empty list of Patterns as its args list. [[pattern.Top]], [[pattern.Bottom]] extend this trait.
+    */
+  sealed trait Node0 extends Node {
+    override def args = Seq.empty[Pattern]
+
+    def build(): Pattern
+
+    override def build(children: Seq[Pattern]): Pattern = {
+      assert(children.isEmpty)
+      build()
+    }
+  }
+
+  object Node0 {
+    def unapply(arg: Node0): Boolean = true
+  }
+
+
+  /**
+    * A Node with a single pattern in its args list. Extended by [[pattern.Next]], [[pattern.Not]].
+    */
+  sealed trait Node1 extends Node with Product1[Pattern] {
+    override def args = Seq(_1)
+
+    def build(_1: Pattern): Pattern
+
+    override def build(children: Seq[Pattern]): Pattern = {
+      assert(children.size == 1)
+      build(children.head)
+    }
+  }
+
+  object Node1 {
+    def unapply(arg: Node1): Option[Pattern] = Some(arg._1)
+  }
+
+
+  /**
+    * A Node with two Patterns in its args list. Extended by [[pattern.Or]], [[pattern.And]], [[pattern.Implies]], [[pattern.Equals]], [[pattern.Rewrite]].
+    */
+  sealed trait Node2 extends Node with Product2[Pattern, Pattern] {
+    def build(_1: Pattern, _2: Pattern): Pattern
+
+    override def args = Seq(_1, _2)
+
+    override def build(children: Seq[Pattern]): Pattern = {
+      assert(children.size == 2)
+      build(children.head, children(1))
+    }
+  }
+
+  object Node2 {
+    def unapply(arg: Node2): Option[(Pattern, Pattern)] = Some(arg._1, arg._2)
+  }
+
+
+  /**
+    * Extends [[Node2]], and only allows [[pattern.Variable]] as the first element, and [[pattern.Pattern]] in its args list. Extended by [[pattern.Exists]], [[pattern.ForAll]].
+    */
+  sealed trait BinderNode extends Node2 {
+    def build(_1: Variable, _2: Pattern): Pattern
+
+    override val _1: Variable
+
+    override def build(_1: Pattern, _2: Pattern): Pattern = build(_1.asInstanceOf[Variable], _2)
+
+    override def build(children: Seq[Pattern]): Pattern = {
+      assert(children.size == 2)
+      build(children.head.asInstanceOf[Variable], children(1))
+    }
+  }
+
+  object BinderNode {
+    def unapply(arg: BinderNode): Option[(Variable, Pattern)] = Some(arg._1, arg._2)
+  }
+
+}
+
+
+object pattern {
+
+  import tree._
+
+  /* ML Pattern Type */
+  sealed trait Pattern extends AST
+
+
+  /**
+    * Matching Logic Variable.
+    *
+    * Provides (Implementations for members)
+    *    - contents of type Product2[Name, Sort].
+    *
+    * Requires (Implementation for members)
+    *    - _1 of type [[Name]].
+    *    - _2 of type [[Sort]].
+    *    - build method taking arguments ([[Name]], [[Sort]]) and returning [[Variable]].
+    */
+  trait Variable extends Pattern with Leaf2[Name, Sort] {
+    def build(_1: Name, _2: Sort): Variable
+  }
+
+  object Variable {
+    def unapply(arg: Variable): Option[(Name, Sort)] = Some(arg._1, arg._2)
+  }
+
+
+  case class Symbol(symbol: String)
```

Comment:
Using str everywhere seems a good way of avoiding confusion. Changed to str in Symbol and Sort